libqtile/core/manager.py

```diff
@@ -422,13 +424,19 @@ def grab_chord(self, chord: KeyChord) -> None:
             hook.fire("enter_chord", chord.mode)
 
         self.ungrab_keys()
+        self.core.grab_keyboard()
         for key in chord.submappings:
-            self.grab_key(key)
+            # TODO figure out how to do this without reaching into backend
+            # internals
+            keysym, mask_key = self.core.lookup_key(key)
+            mask_key &= self.core._valid_mask
+            self.keys_map[(keysym, mask_key)] = key
```

Comment:
I haven't tried it. I switched this out just on general principles, but I don't know what the interaction between GrabKey and GrabKeyboard actually is in X11 protocol.

If that works without problems, it would certainly make the code organization simpler. One potential issue is if the key grabs from GrabKey persist in the X11 state past sending UngrabKeyboard; in that case we would have to either ensure we always called UngrabKey at the same time, or else have problems after the chord was unmapped.

Reply:
I imagine that every time `grab_keyboard` is needed we can safely ungrab and regrab all keys before and after.